docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,324 @@
+
+### Creating the `LazyModule`s
+
+"Lazy" evaluation refers to deferring the evaluation of an expression until
+it is needed.
+
+Parameter negotiation in Diplomacy is done lazily, after the Diplomacy graph
+is created. Because of this, the concrete hardware that we want to parameterize
+must also be generated lazily. Diplomacy provides the `LazyModule` construct for
+writing lazily-evaluated hardware modules.
+
+Let's use `LazyModule` to define our `Adder`. Notice that the creation of the
+components used to define Diplomacy graph (in this case, the node), is non-lazy.
+The desired hardware for the module must be written inside `LazyModuleImp`.
+
+For this example, we expect two drivers, each driving an addend of equal bit
+widths into the adder, meaning we expect all of our downward flowing parameters
+to be equivalent. We also expect a sole monitor, so all of our upward flowing
+parameters should also be equivalent. Thus, we can require these properties in
+our `AdderNode` and simply pass on downward the first `DownwardParam` we see.
+Following the same logic, we can pass upward the first `UpwardParam` we see.
+
+```scala mdoc
+/** adder DUT (nexus) */
+class Adder(implicit p: Parameters) extends LazyModule {
+  val node = new AdderNode (
+    { case dps: Seq[DownwardParam] =>
+      require(dps.forall(dp => dp.width == dps.head.width))
+      dps.head
+    },
+    { case ups: Seq[UpwardParam] =>
+      require(ups.forall(up => up.width == ups.head.width))
+      ups.head
+    }
+  )
+  lazy val module = new LazyModuleImp(this) {
+    require(node.in.size == 2)
+    node.out.head._1 := node.in.head._1 + node.in.tail.head._1
+  }
+
+  override lazy val desiredName = "Adder"
+}
+```
+
+Our `AdderDriver` randomly generates an addend of `width` bits each cycle, and
+passes it to `numOutputs` sources.
+
+```scala mdoc
+/** driver (source)
+  * drives one random number on multiple outputs */
+class AdderDriver(width: Int, numOutputs: Int = 2)(implicit p: Parameters) extends LazyModule {
+  val nodeSeq = Seq.fill(numOutputs)(new AdderDriverNode(DownwardParam(width)))
+
+  lazy val module = new LazyModuleImp(this) {
+    // generate random addend (notice the use of the negotiated width)
+    val randomAddend = FibonacciLFSR.maxPeriod(nodeSeq.head.out.head._2.width)
+
+    // drive signals
+    nodeSeq.foreach { node => node.out.head._1 := randomAddend }
+  }
+
+  override lazy val desiredName = "AdderDriver"
+}
+```
+
+Our `AdderMonitor` prints out the results of the `Adder` each cycle and
+signals an error if the `Adder` returns an incorrect result. It has two
+`AdderMonitorNode`s for taking in addends from the `AdderDriver`, and one
+`AdderMonitorNode` to take in the sum from the `Adder`.
+
+```scala mdoc
+/** monitor (sink) */
+class AdderMonitor(width: Int)(implicit p: Parameters) extends LazyModule {
+  val nodeSeq = Seq.fill(2) { new AdderMonitorNode(UpwardParam(width)) }
```

Comment:
Here "2" should be equivalent to the number of driver modules, so I think you are right to manually create one node for each one of them, expecting one input edge each. Basically, we are saying "every Adder Monitor rightly expects that there are exactly 2 input numbers each coming from a different driver." But contrast this comment to what I am about to say for `AdderDriverNode` above

Reply:
But maybe there should it should also be an argument like `numOperands`?

Reply:
Hmm I guess youd have to make the below logic more complicated too, so nevermind for now

Reply:
I think it does look nicer to generalize all the modules to `n` args, so I think I can finagle this one, too!